src/Features/ExternalAccess/HotReload/Api/HotReloadMSBuildWorkspace.cs

```diff
@@ -0,0 +1,232 @@
+            {
+                documentsToRemove.AddRange(documentIds);
+                continue;
+            }
+
+            foreach (var documentId in documentIds)
+            {
+                var textDocument = updatedSolution.GetDocument(documentId)
+                    ?? updatedSolution.GetAdditionalDocument(documentId)
+                    ?? updatedSolution.GetAnalyzerConfigDocument(documentId);
+
+                if (textDocument == null)
+                {
+                    _logger.LogDebug("Could not find document with path '{FilePath}' in the workspace.", path);
+                    continue;
+                }
+
+                var project = updatedSolution.GetProject(documentId.ProjectId);
+                Debug.Assert(project?.FilePath != null);
+
+                var oldText = await textDocument.GetTextAsync(cancellationToken).ConfigureAwait(false);
+                Debug.Assert(oldText.Encoding != null);
+
+                var newText = await GetSourceTextAsync(path, oldText.Encoding, oldText.ChecksumAlgorithm, cancellationToken).ConfigureAwait(false);
+
+                _logger.LogDebug("Updating document text of '{FilePath}'.", path);
+
+                updatedSolution = textDocument switch
+                {
+                    Document document => document.WithText(newText).Project.Solution,
+                    AdditionalDocument ad => updatedSolution.WithAdditionalDocumentText(textDocument.Id, newText, PreservationMode.PreserveValue),
+                    AnalyzerConfigDocument acd => updatedSolution.WithAnalyzerConfigDocumentText(textDocument.Id, newText, PreservationMode.PreserveValue),
+                    _ => throw ExceptionUtilities.UnexpectedValue(textDocument),
+                };
+            }
+        }
+
+        updatedSolution = RemoveDocuments(updatedSolution, documentsToRemove);
+
+        return SetCurrentSolution(updatedSolution);
+    }
+
+    private static Solution RemoveDocuments(Solution solution, IEnumerable<DocumentId> ids)
+        => solution
+        .RemoveDocuments([.. ids.Where(id => solution.GetDocument(id) != null)])
+        .RemoveAdditionalDocuments([.. ids.Where(id => solution.GetAdditionalDocument(id) != null)])
+        .RemoveAnalyzerConfigDocuments([.. ids.Where(id => solution.GetAnalyzerConfigDocument(id) != null)]);
+
+    private static async ValueTask<SourceText> GetSourceTextAsync(string filePath, Encoding encoding, SourceHashAlgorithm checksumAlgorithm, CancellationToken cancellationToken)
+    {
+        var zeroLengthRetryPerformed = false;
+        for (var attemptIndex = 0; attemptIndex < 6; attemptIndex++)
+        {
+            try
+            {
+                // File.OpenRead opens the file with FileShare.Read. This may prevent IDEs from saving file
+                // contents to disk
+                SourceText sourceText;
+                using (var stream = File.Open(filePath, FileMode.Open, FileAccess.Read, FileShare.ReadWrite))
```

Comment:
Our regular implementation for reading files is also passing FileShare.Delete:

https://github.com/dotnet/roslyn/blob/e69fc360fc88a5374547e3d88f6f2c6c11aac9f9/src/Workspaces/Core/Portable/Workspace/Solution/FileTextLoader.cs#L94